cli/internal/git/utils.go

```diff
@@ -151,3 +159,155 @@ func CredentialsGenerator() string {
 
 	return gitSecret
 }
+
+// GetTaggedUrl builds a URL of the repo@tag format
+// It returns a string of format repo@tag
+func GetTaggedUrl(gitUrl string, gitTag string) string {
+	return gitUrl + "@" + gitTag
+}
+
+// RemoveLocalBranchRefs removes all refs that are local branches
+// It returns a slice of references deleted
+func RemoveLocalBranchRefs(gitDirectory string) []*plumbing.Reference {
+	return removeReferences(
+		gitDirectory,
+		func(ref *plumbing.Reference) bool {
+			return ref.Name().IsBranch()
+		},
+	)
+}
+
+// RemoveOnlineRemoteRefs removes all refs pointing to the online-upstream
+// It returns a slice of references deleted
+func RemoveOnlineRemoteRefs(gitDirectory string) []*plumbing.Reference {
+	return removeReferences(
+		gitDirectory,
+		func(ref *plumbing.Reference) bool {
+			return strings.HasPrefix(ref.Name().String(), onlineRemoteRefPrefix)
+		},
+	)
+}
+
+// RemoveHeadCopies removes any refs that aren't HEAD but have the same hash
+// It returns a slice of references deleted
+func RemoveHeadCopies(gitDirectory string) []*plumbing.Reference {
+	logContext := logrus.WithField("Repo", gitDirectory)
+	repo, err := git.PlainOpen(gitDirectory)
+	if err != nil {
+		logContext.Debug(err)
+		logContext.Fatal("Not a valid git repo or unable to open")
+	}
+
+	head, err := repo.Head()
+	if err != nil {
+		logContext.Debug(err)
+		logContext.Fatal("Failed to identify references")
+	}
+
+	headHash := head.Hash().String()
+	return removeReferences(
+		gitDirectory,
+		func(ref *plumbing.Reference) bool {
+			// Don't ever remove tags
+			return !ref.Name().IsTag() && ref.Hash().String() == headHash
+		},
+	)
+}
+
+// removeReferences removes references based on a provided callback
+// removeReferences does not allow you to delete HEAD
+// It returns a slice of references deleted
+func removeReferences(
+	gitDirectory string,
+	shouldRemove func(*plumbing.Reference) bool,
+) []*plumbing.Reference {
+	logContext := logrus.WithField("Repo", gitDirectory)
+	repo, err := git.PlainOpen(gitDirectory)
+	if err != nil {
+		logContext.Debug(err)
+		logContext.Fatal("Not a valid git repo or unable to open")
+	}
+
+	references, err := repo.References()
+	if err != nil {
+		logContext.Debug(err)
+		logContext.Fatal("Failed to identify references")
+	}
+
+	head, err := repo.Head()
+	if err != nil {
+		logContext.Debug(err)
+		logContext.Fatal("Failed to identify head")
+	}
+
+	removedRefs := []*plumbing.Reference{}
+	err = references.ForEach(func(ref *plumbing.Reference) error {
+		passHead := !(ref.Name() == plumbing.HEAD || ref.Name() == head.Name())
```

Comment:
Not super clear on this, the negative logic is a little awkward, can we do `ref.Name() != plumbing.HEAD && ref.Name() != head.Name()` instead?  The name passHead isn't very clear either.  Maybe a little more comments on what is happening here?

Reply:
Sure! 

To explain, I called it `passHead` because it determines if we pass the HEAD check **for deletion**, which is `and`ed with if we pass the user check for deletion. I can find a better name for it though since `passHead` certainly isn't very descriptive on its own. As for the awkward logic, I wrote it that way to reflect "what would a HEAD check look like", and then "now invert that since we want to pass the HEAD check when the ref is _not_ HEAD or what HEAD points to". Refactoring to "what implies that the ref isn't HEAD or what HEAD points to" makes sense to me though. I'll add comments when I do so as well.